### PR TITLE
allow empty secret in gcs BR

### LINKED ISSUE #3557: Support GCP service account when running backup/restore on GCP
## Feature Request

**Is your feature request related to a problem? Please describe:**

When running backup/restore on GCP, we have to create a secret containing the credentials, otherwise the Backup would fail to create.

**Describe the feature you'd like:**

With BR support GCP application default credentials (ADC) https://github.com/pingcap/br/issues/633, tidb-operator should support empty secret for GCS backup/restore.

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

https://cloud.google.com/docs/authentication/production

### PR DESCRIPTION
**Pull request by BinChenn**
### What problem does this PR solve?
#3557. 

Allow empty secret when backup/restore to/from gcs in BR.

### What is changed and how does it work?
Change `GcsStorageProvider.SecretName` to`omitempty`, add UT.

### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [x] Unit test 

- [x] E2E test 

- [x] Manual test 

    - Backup to gcs with BR configuring SecretName

        1. Create a kind cluster and deployment tidb cluster

            ```shell
            kind create cluster
            ./hack/local-up-operator.sh
            kubectl apply -f example/basic/tidb-cluster.yaml
            ```

        2. [Backup to gcs with BR](https://docs.pingcap.com/zh/tidb-in-kubernetes/stable/backup-to-gcs-using-br)

            ```yaml
            curl https://raw.githubusercontent.com/pingcap/tidb-operator/master/manifests/backup/backup-rbac.yaml
            kubectl apply -f backup-rbac.yaml -n test1
            kubectl create secret generic gcs-secret --from-file=credentials=gcs-credentials.json
            kubectl create secret generic backup-demo1-tidb-secret --from-literal=password=${password}
            ```

            backup-gcs.yaml

            ```yaml
            ---
            apiVersion: pingcap.com/v1alpha1
            kind: Backup
            metadata:
              name: demo1-backup-gcs
              namespace: default
            spec:
              from:
                host: basic-tidb.default
                port: 4000
                user: root
                secretName: backup-demo1-tidb-secret
              br:
                cluster: basic
                clusterNamespace: default
              gcs:
                projectId: ${project-id}
                secretName: gcs-secret
                bucket: ${bucket}
                prefix: ${backup}
            ```

            ```shell
            kubectl apply -f backup-gcs.yaml
            ```

            Backup success

            ```
            I1211 09:47:38.339247       1 backup.go:107] Backup data for cluster default/demo1-backup-gcs successfully
            I1211 09:47:38.349994       1 manager.go:274] reset cluster default/demo1-backup-gcs tikv_gc_life_time to 10m0s success
            I1211 09:47:38.350022       1 manager.go:289] backup cluster default/demo1-backup-gcs data to gcs://chenbin_test/backup success
            ```

    - [TODO]Backup to gcs with BR configuring empty SecretName

        Waiting for TiKV [support](https://github.com/pingcap/br/issues/633) GCP application default credentials.

- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
NONE
```

### CODE REVIEW

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 1]
```diff
@@ -1197,7 +1197,7 @@ type GcsStorageProvider struct {
 	BucketAcl string `json:"bucketAcl,omitempty"`
 	// SecretName is the name of secret which stores the
 	// gcs service account credentials JSON.
-	SecretName string `json:"secretName"`
+	SecretName *string `json:"secretName,omitempty"`
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1200]: I think just a string is OK, no need to be a pointer.

Comment (BinChenn) [reply pkg/apis/pingcap/v1alpha1/types.go:1200]: fixed

File: pkg/backup/testutils/br.go [hunk 2]
```diff
@@ -51,6 +51,13 @@ func GenValidStorageProviders() []v1alpha1.StorageProvider {
 				SecretName: "gcs",
 			},
 		},
+		{
```
Comment (DanielZhangQD) [inline pkg/backup/testutils/br.go:54]: What's the difference with that in L40-L44?

Comment (BinChenn) [reply pkg/backup/testutils/br.go]: Unnecessary test case, removed!

**Comment by DanielZhangQD** [comment]
@BinChenn If `gcs.secret` is not configured, we have to make sure that the backup and restore Jobs are created successfully but only the real jobs will fail.

**Comment by BinChenn** [comment]
> @BinChenn If `gcs.secret` is not configured, we have to make sure that the backup and restore Jobs are created successfully but only the real jobs will fail.

Test done. When `gcs.secret` is not configured, jobs could be created successfully, then real br jobs will fail.

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Comment by handlerww** [comment]
LGTM

**Review by handlerww — APPROVED** [review]

**Review by tennix — APPROVED** [review]
LGTM

**Pull request merged**